Refactor code to check permissions for updating code.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit is a refactor commit so that we can use different settings for checking permissions for different things like updating group, adding members, etc.
- Second commit is to add restriction for allowing only group members for managing group.

<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented:
I tried this refactoring so that we can easily use different settings to check permissions for different things like managing group, adding members to group, etc. without needing to write duplicate code, especially the group member check for non-admin and non-moderator users.

sahil839 commented:
Tests might fail, but I can fix this once I get some initial feedback on the approach and whether such a refactoring would help. @shubham-padia can you take a look?

shubham-padia commented:
The general approach in the first commit looks good, I haven't dived deeper, let me know when it's ready for that. Also, haven't reviewed the second commit since it may not be relevant after https://chat.zulip.org/#narrow/stream/3-backend/topic/Group.20management.20-.20admin.20checks/near/1949081

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -78,8 +78,6 @@ def has_user_group_access(
     user_profile: UserProfile,
     *,
-    for_read: bool,
-    as_subgroup: bool,
```
`has_user_group_access`  is only used for verifying subgroup access and it only checks deactivated check currently.

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -88,66 +86,67 @@ def has_user_group_access(
     *,
-    for_read: bool,
-    for_setting: bool = False,
+    permission_setting: str,
```
This needs some hardening, maybe using a `Literal` type, again typos in the caller of the setting name. Or maybe doing a `getattr(user_group, permission_setting)` earlier in the function to confirm it's a valid property would be cleaner?

sahil839 replied:
How about using `assert permission_setting in NamedUserGroup.GROUP_PERMISSION_SETTINGS`.

timabbott replied:
Yes that sounds good.

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -88,66 +86,67 @@ def has_user_group_access(
-    )
+        group_members = get_user_group_direct_member_ids(user_group)
+        if user_profile.id in group_members:
+            return True
```
This requires fetching all 20k users if it's a big group; we should just check if the user is in the group with the function for that.

sahil839 replied:
Yes, we should do that. Should we also change this to consider members of recursive subgroups as well instead of just allowing direct members.

timabbott replied:
Yes, of course.

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -404,8 +405,8 @@ def access_user_group_for_setting(
-        named_user_group = access_user_group_by_id(
-            setting_user_group, user_profile, for_read=True, for_setting=True
+        named_user_group = get_user_group_by_id_in_realm(
+            setting_user_group, user_profile.realm, for_read=False
```
Why is this one switching to `for_read=False`?

sahil839 replied:
We previously had `for_setting` parameter which was passed as `True` to make sure that we lock the user group row, which is now acheived by passing `for_read` as `False`. I actually had to ask this before, do we need a separate `for_setting` parameter and pass `for_read=True`? We actually do not update the group fetched here and just use it for a setting, so we are just reading the object.

timabbott replied:
Yes, it's important. I believe the intent is to lock the group to prevent it from being deleted here, as part of our controls to avoid having a deleted/deactivated group ever being used as a setting value.

sahil839 replied:
I was passing `for_read=False` so that the group is locked. But yeah I have just added `for_setting` parameter again now since that felt more clear.

timabbott commented:
The refactor makes sense aside from the above comments on naming. But I'm a bit worried in the discussion of a second commit -- I'm not aware of settings where we want group membership to impact how the permission works once we've cleaned up and migrated the legacy `edit_group_policy` logic.

sahil839 commented:
I am also confused about how we want the group membership to impact the permissions.

For `can_manage_all_groups`, which will replace `user_group_edit_policy`, we allow members only if they are part of the group, which makes sense for keeping the logic same when migrating from old setting to new. But do we want to keep this restriction forever or just remove it after some time?

And similarly for "Who can join this group" setting, currently the permission is controlled by `user_group_edit_policy`. We would add this setting after adding `can_manage_all_groups`, since we probably plan to keep the default of this setting to the value set for `can_manage_all_groups` setting. So, how we want to handle the value of "Who can join this group" setting for existing realms? We would need to have the restriction of allowing members only if they are part of the group.

sahil839 commented:
A workaround might be for setting `can_manage_all_groups` value to "Admins and moderators" for existing realms that have the value set to "Admins, moderators and members" or "Admins, moderators and full members" and make it clear that members would have control over all the user groups through that setting irrespective of the membership.

timabbott commented:
I'll post about this in chat.zulip.org as well, but I think compatibility with existing values of `user_group_edit_policy` is not a goal or requirement for our updated settings; prior to this transition, user groups provided no permissions, and the existing values for that setting are more or less unrelated to what values we want for the new setting.

I expect we're going to default to a fixed value like administrators for `can_manage_all_groups`, and these messy checks about group membership will be replaced with (1) initial values for "who can add others to this group" settings attached to individual groups -- to be finalized when working on that PR and (2) potentially a future check for group access, if we make some groups not visible to all users in the future.

I think we need to design a focused complete set of transition steps that will make this easier to organize over the next couple weeks, will post in chat.zulip.org about it.

sahil839 commented:
Updated the PR for changes mentioned above. Also added a commit to allow members of subgroups of a group, while checking the permission for managing groups.

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -74,7 +74,7 @@ class LockedUserGroupContext:
 
 
-def has_user_group_access(
+def has_user_group_access_for_subgroup(
```
It's weird that this sometimes throws exceptions.

timabbott commented:
This looks reasonable; I pushed a bonus commit adding docstrings. 

I think `get_user_group_by_id_in_realm` would ideally be an internal function only used in this module, and the `zerver/views/user_groups` code could call some wrapper, like `access_user_group_to_read_membership`, for better abstractions; it's not a safe function to call with `for_read=False` from client code, since it doesn't actually do any security checks. Can you look at that as a follow-up?
